### PR TITLE
Argument Guesser analysis.

### PR DESCRIPTION
**Pull request by uxmal**
This new analysis does a reasonably good job in guessing the actual arguments passed to procedures whose signaturs are unknown. Currently, it's limited to `ExternalProcedure`s. That is, procedures that have been imported from DLL's/shared libraries, but whose signatures Reko is unaware of, for lack of metadata. 

The strategy is simple: walk backwards from call sites of `ExternaProcedure`s and collect `Assignment`s that look like they are part of argument passing protocol. This means using the ABI input and output registers, and looking at assignments to stack locations. The backwards walk stops when instructions that are not related to argument setups are encountered (like `PhiAssignment`s or `CallInstruction`s) 

Because this is a speculative analysis, it will generate incorrect results in some cases. Therefore, more work needs to be done to make this analysis optional (on a whole-program and perhaps on a per-procedure basis), and also to allow the user to review the guesses and improve the guesses manually if desired.

### CODE REVIEW

**Review by ptomin — COMMENTED** [review]
Almost approved. Just problem with `ssa id` uses and some minor notes.

File: src/Decompiler/Analysis/ArgumentGuesser.cs [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+                newInstr = new Assignment(idRet, application);
+            }
+            else
+            {
+                newInstr = new SideEffect(application);
+            }
+            ssa.RemoveUses(stmCall);
+            ssa.ReplaceDefinitions(stmCall, null);
+            stmCall.Instruction = newInstr;
+            var ssam = new SsaMutator(this.ssa);
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Decompiler/Analysis/ArgumentGuesser.cs:298]: Dead code removed.

File: src/Decompiler/Analysis/ArgumentGuesser.cs [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+                {
+                    args.Add(stkarg.Identifier);
+                }
+                foreach (var stackslot in gargs.Value.StackSlots.Values.OrderBy(s => s.Offset))
+                {
+                    ssa.RemoveUses(stackslot.stm);
+                    ssa.Identifiers[stackslot.Dst.MemoryId].DefStatement = null;
+                    var idTmp = binder.CreateTemporary(stackslot.Dst.DataType);
+                    var sidTmp = ssa.Identifiers.Add(idTmp, stackslot.stm, stackslot.src, false);
+                    stackslot.stm.Instruction = new Assignment(sidTmp.Identifier, stackslot.src);
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Decompiler/Analysis/ArgumentGuesser.cs:280]: We're replacing 
```
Mem[sp - NN:wordXX] = <src>
```
with
```
tmpNN = <src>
```
The uses of `<src>` remain the same.

**Comment by uxmal** [comment]
Ready to go? A future commit will make the `ArgumentGuesser` optional and add support for manual overriding of external procedures' metadata.

**Review by ptomin — APPROVED** [review]
Looks good now.

**Pull request merged**